### ISSUE TITLE
how to use loaded db outside of callback functions

### ISSUE THREAD

**Issue by tommyhtc** [techfort/LokiJS#108]
hi there, so I'm just learning loki,

let's say I have a loki database called "test.json" with a collection "Users".
I'm running node,
In some .js file, I want to access that database and write to it.

Currently, I only know how to access a loaded database with callbacks,
for example, with:

db = new loki('test.json',{autoload:true, autocallback: foo()});

and then I'm able to access the loaded db in foo(), where I can insert stuff or get stuff, etc.

However, is there a way to access the loaded database outside of the callback, or without using callback??

for example,
foo(){
   db = new loki('test.json',{autoload:true}
   somefield = db.getCollection("Users").get(1);
}

is this possible??
I've tried loadDatabase but it also requires a callback..

**Comment by obeliskos** [comment]
Yea, your callback can be anonymous such as : 
```javascript
var db = new loki('test.json');
db.loadDatabase({}, function() {

});
```
Or you could edit your copy of loki.js and change : 
```javascript
    LokiFsAdapter.prototype.loadDatabase = function loadDatabase(dbname, callback) {
      var data = this.fs.readFileSync(dbname, {
        encoding: 'utf8'
      });
      callback(data);
    };

```

This seems academic, is this something you see others wanting to do? Is it worth allowing user to configure an option to force synchronicity?

**Comment by techfort** [comment]
I have to look into the details of it but you could implement the load as a promise. Ultimately I/O is async so there's going to be some form of callback involved, whether a normal node.js style one a promise-style one.

**Comment by tommyhtc** [comment]
Hi guys, thanks for your responses..
while I'm new to loki, I'm also new to node.js,

it's only after I asked the question did I learn about the asynchronous nature of node.js,

this is probably a basic question,
but let's say I have two functions create_person and create_edge where create_edge creates an edge between two people by storing two person ids.

for both create_person and create_edge, I use callbacks to write to a database.

if I call create_person("A"), create_person("B"), then create_edge("A","B").
by the time I call create_edge("A","B"), would it be guaranteed that the callbacks of create_person("A") and create_person("B") are already called and that both person records have been added to the database?

**Comment by obeliskos** [comment]
The asynchronous nature is mostly limited to I/O operations... for loki this is mostly loading and saving.
Inserts, updates, removes are synchronous in loki, so once you call them you don't need a callback, they are done.

Your example is more representative of a graph database.  Are you doing this for school or personal?  If school, you may want to research graph databases which work in node.

**Comment by tommyhtc** [comment]
hey, this is for a school project, for various reasons our team decided to use loki.

when I say callback, I think I may have misled you, I'm referring to the anonymous functions that you talked about above, 
for example, foo() in this case, in all of my insert functions, I do the inserts in the anonymous function.
When I'm calling insert, I'm not calling any other callbacks.

I'm only saying I'm calling insert in the callbacks from loading the database.

var db = new loki('test.json');
db.loadDatabase({}, foo() {
  somefield = db.getCollection("Users").get(1);
});

so when you say the asynchronous nature is limited to I/O operations, does that mean the scenario I mentioned above won't be a problem?? 
so as I've said, if I call functionA(), functionB(), and functionC().
if each function loads the database, which invokes a callback that does the insertion.

So by the time the callback function for insertion in functionC() is called, will the callbacks of functionA() and functionB() be executed??

**Comment by techfort** [comment]
@tommyhtc ok - let's take a step back. JS is single-threaded. This means things cannot EVER happen concurrently (academically speaking this does not even happen in languages that only support threading, but only in those supporting concurrency, but that's a whole other story). 
So, talking about normal code, if you have:
```javascript
function a() { return 1; }
function b() { return 2; }
function c() { return 3; }
// and call
a();
b();
c();
```
they will execute in order, no matter what.
The asynchronicity problem enters when you deal with one of these:
* I/O (including network commuincations, e.g. ajax, or http requests on server-side)
* you use `setInterval` or `setTimeout`

Async in javascript means this (roughly): as soon as the existing sync code has been executed, if the processor is free to do other things, then i'll execute async.
For example, you may set a timeout for a function to execute after 1ms, let's say prompt an alert. The you also stick in a function that does heavy processing, like this:
```javascript
setTimeout(function () { alert('hello'); }, 1);
heavyProcessingStuff();
```
You instruct to execute that alert after 1ms, but heavyProcessingStuff() kicks in before that, so now the processor has to wait for heavyProcessingStuff() to be done before it can execute it. And if heavyProcessingStuff() takes 35 minutes, well that's what you're going to have to wait.

In conclusion: LokiJS adopts a synchronous style of programming, except for when it can't due to node.js's own event-loop, namely disk or network I/O. The only circumstance when you will need to pay attention to callbacks are loads and saves to disk, and - if you do it- syncing to remote servers through requests (like ajax or server-side http() ). Other than that ... you're good.

**Comment by obeliskos** [comment]
Great explanation @techfort 

What you are describing is a 'graph' database.  A graph database is one type of NOSQL database.  LokiJS is primarily a 'document-oriented' database.  You can however try to simulate a graph database within LokiJS, I just feel I need to mention this in case your instructor cares.  Eventually loki make include relations which may make this easier, for now you have to relate manually with object id's.

I would guess you want a collection for Users and a collection for Vertex.  The vertex collection might hold objects similar to : 
```javascript
var db = new loki("test.db");
var userColl = db.addCollection("users");

var user = { 'first': 'john', 'last': 'doe' }
var user2 = { 'first': 'jane', 'last': 'doe' }
userColl.insert(user);
userColl.insert(user2);

var vertColl = db.addCollection("vertices");
// $loki is primary key automatically added to an object when inserted
var vertex { u1 : user.$loki, u2 : user2.$loki, desc:'is married to' }  
vertColl.insert(vertex);

// now lookup objects linked to first user and log relation name and other user linked to it
var result = vertColl.find({'u1': user.$loki});
console.log('users related to : ' + user.first + ' ' + user.last + ' : ');
var tmpUser;
result.foreach(function(obj)) {
  console.log(obj.desc);
  tmpUser = userColl.get(obj.u2);
  console.log(tmpUser.first + ' ' + tmpUser.last);
}
```
You might also read up on databases like Neo4J to get an idea of what you are trying to simulate.

**Comment by tommyhtc** [comment]
guys, I appreciate your thoughtful and useful responses, it's a great learning process for me.

@obeliskos, thanks again for your suggestions. We are doing something very similar to what you had in the code snippet, simulating a graph database using loki. And while graph databases have come up in our discussions, our group decided not to use a graph database for now.

@techfort 
thanks again, I appreciate your response. It is crystal clear and very useful.

I do understand that for normal functions sequence like the one you used in your explanation,
they will execute in order. This makes sense and is intuitive, because it's like any other popular programming language.

but I want to make sure I completely understand what you mean when it comes to I/O stuff.
I feel like this is probably not a specific loki question, but may be more like a general cs/javascript/node question.

if I have functions like these:
function A(){
   var db = new loki('test.json',{autoload:true,autocallback: Acallback()});
   function Acallback(){
      // save something to test.json using loki.save()
   }
}
function B(){
     var db = new loki('test.json',{autoload:true,autocallback: Bcallback()});
   function Bcallback(){
      // save something to test.json using loki
   }
}

if I call
functionA();
functionB();

by the time I call functionB(), everything from functionA(), including it's callback would have already terminated, right?? or is this exactly what you say I should be careful of???

**Comment by techfort** [comment]
hehe no :)
node.js in a way "washes its hands" of what happens while I/O is executing. In theory, given that A and B deal with the same file, Acallback should execute before Bcallback, but there is no guarantee of that. If the read operation encounters problems, it may well be that Acallback will end up being executed after Bcallback.
So yes, this is exactly what you need to be careful of.
If you need - at all costs - that Bcallback be executed after Acallback then I suggested you call B() inside Acallback, that way the whole B() logic will start only when A()'s I/O is finished.
good luck!

**Comment by obeliskos** [comment]
I/O is input/output... anything from reading/writing files, receiving/transmitting web requests/responses, etc.  I/O is generally -much- slower so to avoid having the cpu lock up your program waiting for responses async allows you to continue.

You should not be creating loki in each function, maybe do something like : 
```
var db = new loki('test.json',{autoload:true,autocallback: dbLoaded()});
var userColl = db.addCollection('users');

function dbLoaded() 
{
    functionA();
    functionB();
    functionC();

    // can save database here if needed or configure autosave in first line of example
    db.saveDatabase();
}

function A() {
   userColl.insert({first: 'john', last: 'doe'});   
}

function B() {
   userColl.insert({first: 'john', last: 'doe'});   
}

function C() {
   userColl.insert({first: 'another', last: 'user'});   
}

**Comment by techfort** [comment]
@obeliskos thanks for showing the pattern, that should help a lot @tommyhtc

**Comment by tommyhtc** [comment]
@techfort , @obeliskos 
many thanks for your help.
Now everything is clear and I can go modify my code now!

**Comment by obeliskos** [comment]
Glad to help, good luck

**Comment by joshuaquek** [comment]
I built a library that might address this issue:

https://www.npmjs.com/package/lokijs-promise

Decided to build this as quite a lot of the issues that I found in the Issues section revolved around loading the persistent store (eg. `store.json`) and then ensuring that the store was loaded before doing inserts etc, which then automatically writes to the persistent store (eg. `store.json`)
